### ISSUE TITLE
AnglE loss reimplementation

### ISSUE THREAD

**Issue by user799595** [UKPLab/sentence-transformers#3223]
I spent some time understanding the AnglE loss function, specifically, `pairwise_angle_sim` and have come up with a cleaner implementation.

Would this be of interest?

**Comment by tomaarsen** [comment]
Hello!

Certainly, if the new function is clearer/cleaner, and especially if it is faster, then I'm very interested. I do want to make sure that it performs the same.

- Tom Aarsen

**Comment by user799595** [comment]
Hi Tom,

I'm sending the description for now, code to follow. I have tested the correctness of the code on random inputs, but not the performance. 

AnglE similarity is a modification of cosine similarity with an addition "sine similarity" term. To define sine similarity, the authors pair up dimensions. If the total embedding dimension is D, each dimension i < D/2 is paired with i+D/2. Angle similarity is not defined for odd D.

For normalized vectors, similarity increases when v[i] becomes more similar w[i].
Sine similarity increases when v[i] becomes similar to w[i+D/2] and v[i+D/2] becomes similar to -w[i] for i < D/2. Note the minus sign. Its full definition is:
```sin_sim(v, w) = cos_sim(v, torch.cat(-w[D/2:], w[:D/2])```

The minus sign implies that sin_sim(v,w) = - sin_sim(w, v), so order matters. The order of dimensions also matters, because of the pairing. So for a random permutation p,

```cos_sim(v[p], w[p]) = cos_sim(v, w)```

This will not in general be the case for sine similarity. It assumes a special linkage between dimension pairs.

The final definition is

```angle_sim(v, w) = abs(cos_sim(v, w) + sin_sim(v, w))```

**Comment by user799595** [comment]
```python
def pairwise_angle_sim2(x: Tensor, y: Tensor) -> Tensor: 
    normalized_x = torch.nn.functional.normalize(x, p=2, dim=1)
    normalized_y = torch.nn.functional.normalize(y, p=2, dim=1)
    cosines = torch.sum(normalized_x * normalized_y, dim=1)
    first_x, last_x = torch.chunk(normalized_x, 2, dim=1)
    first_y, last_y = torch.chunk(normalized_y, 2, dim=1)
    sines = torch.sum(last_x * first_y - first_x * last_y, dim=1)
    return torch.abs(cosines+sines)
```

**Comment by johneckberg** [comment]
Hi all! 

I may be able to provide some helpful feedback as the one who implemented `pairwise_angle_sim` in Sentence Transformers (based on the original implementation by [SeanLee97](https://github.com/SeanLee97/AnglE/blob/main/angle_emb/angle.py#L114)). 

@user799595, could you clarify what you mean by "an additional sine similarity" term? My understanding is that angle similarity from the [AnglE paper](https://aclanthology.org/2024.acl-long.101.pdf) is not cosine similarity with an added term; rather it is the element-wise sum of the absolute angle difference between two complex embeddings in polar form. 

The first step of angle similarity is to transform the Bert output into one real and one imaginary vector. This is a process inspired by [RotatE](https://openreview.net/forum?id=HkgEQnRqYQ). We take the first half of the vector as our real component and the second half as the imaginary component, giving us the two complex embeddings mentioned above (Z,W). The Angle difference, 𝜃<sub>zw</sub>, refers to Z/W divided by a normalizing constant in polar form, rather than the angle between our pair sentence vectors from the embedding space.  

While there are bijections between vector spaces over the complex numbers and vector spaces over the real numbers (C<sup>n</sup> ≅ R<sup>2n</sup>), this loss function deals with objects that instead act like vector valued complex numbers. Instead a vector in complex Euclidean space C<sup>n</sup>, we have a 2D complex plane with where each point is described by two high dimensional (R<sup>2n</sup>) real vectors. This is pretty un-intuitive, so for an example, consider a vector V in the space R<sup>6</sup>. This vector has exactly one point it corresponds to in C<sup>3</sup>, and vice versa because of the bijection between the vector spaces. Since every dimension  in C<sup>3</sup> is complex, each dimension of each point could also be represented in polar form. 

However, instead, imagine the same vector V in the same space R<sup>6</sup>. What `pairwise_angle_sim` does is split that input vector into 2 parts, rather than splitting the input vector up into 3 (or d/2) parts like in the previous example. Say we have two real vectors in either of the "saturation zones", meaning the angle 𝜃 between them is very large or very small.  Recall that there's a one to one and onto relationship between (C<sup>n</sup> and R<sup>2n</sup>). If 𝜃 is the angle between two points in R<sup>2n</sup> it is also unfortunately the angle between their mirror images in C<sup>n</sup>. This means our gradient still vanishes. To prevent this, we instead find the absolute angle difference between Z and W, which provides a stronger learning signal in the saturation zones. Then we pool by summation as the angle difference is a vector in polar form. This is a different pooling step done than the one done by an sentence embedding model to get the pooled text embedding vector. 

You are correct in saying that this process is not well defined for an odd embedding dimension, but I'm afraid that is unavoidable due to the nature of how these complex embeddings are created in the [AnglE paper](https://aclanthology.org/2024.acl-long.101.pdf). Looking at the code for `pairwise_angle_sim` now, I definitely think that variable names could improved for clarity. It is a bit of an unusual technique and is hard to follow. Further, it looks like Sean added a way to pool besides summation in his [repo](https://github.com/SeanLee97/AnglE/blob/main/angle_emb/angle.py#L114), which is a feature that could be added to the Sentence Transformers implementation. Definitely improvements to be made!

Always happy to answer any questions!
John

**Comment by user799595** [comment]
Hi @johneckberg 

Thanks for chiming in. I tested my implementation on some randomly drawn tensors and it matches `pairwise_angle_sim` perfectly, so I think we're discussing two different interpretations of the same function.

I thought above that I would share an alternative implementation/interpretation of the method, in case people find it useful.

I found the paper's interpretation quite difficult to follow, and in truth, I read your message several times and I'm still not sure I get 100% of the content. For instance, it's not clear to me from the complex numbers interpretations that ```pairwise_angle_sim(v, w) != pairwise_angle_sim(w, v)```

**Comment by johneckberg** [comment]
I apologize, I think I may be misunderstanding your original comments. Happy to hear that your alternative implementation is testing correctly!
